routing/cross_mwm_graph.cpp

```diff
@@ -0,0 +1,201 @@
+#include "routing/cross_mwm_graph.hpp"
+
+#include "indexer/scales.hpp"
+
+#include "geometry/mercator.hpp"
+
+#include "base/stl_helpers.hpp"
+
+#include "defines.hpp"
+
+#include <algorithm>
+#include <numeric>
+#include <utility>
+
+using namespace routing;
+using namespace std;
+
+namespace
+{
+double constexpr kTransitionEqualityDistM = 20.0;
+double constexpr kInvalidDistance = numeric_limits<double>::max();
+
+struct ClosestSegment
+{
+  ClosestSegment() = default;
+  ClosestSegment(double minDistM, Segment const & bestSeg, bool exactMatchFound)
+    : m_bestDistM(minDistM), m_bestSeg(bestSeg), m_exactMatchFound(exactMatchFound)
+  {
+  }
+
+  double m_bestDistM = kInvalidDistance;
+  Segment m_bestSeg;
+  bool m_exactMatchFound = false;
+};
+}  // namespace
+
+namespace routing
+{
+CrossMwmGraph::CrossMwmGraph(Index & index, shared_ptr<NumMwmIds> numMwmIds,
+                             shared_ptr<VehicleModelFactory> vehicleModelFactory,
+                             RoutingIndexManager & indexManager)
+  : m_index(index)
+  , m_numMwmIds(numMwmIds)
+  , m_vehicleModelFactory(vehicleModelFactory)
+  , m_crossMwmIndexGraph(index, numMwmIds)
+  , m_crossMwmOsrmGraph(numMwmIds, indexManager)
+{
+  CHECK(m_numMwmIds, ());
+  CHECK(m_vehicleModelFactory, ());
+}
+
+bool CrossMwmGraph::IsTransition(Segment const & s, bool isOutgoing)
+{
+  // Index graph based cross-mwm information.
+  if (CrossMwmSectionExists(s.GetMwmId()))
+    return m_crossMwmIndexGraph.IsTransition(s, isOutgoing);
+  return m_crossMwmOsrmGraph.IsTransition(s, isOutgoing);
+}
+
+void CrossMwmGraph::GetTwins(Segment const & s, bool isOutgoing, vector<Segment> & twins)
+{
+  CHECK(IsTransition(s, isOutgoing),
+        ("The segment", s, "is not a transition segment for isOutgoing ==", isOutgoing));
+  // Note. There's an extremely rare case when a segment is ingoing and outgoing at the same time.
+  // |twins| is not filled for such cases. For details please see a note in
+  // CrossMwmGraph::GetEdgeList().
+  if (IsTransition(s, !isOutgoing))
+    return;
+
+  twins.clear();
+
+  // Note. The code below looks for twins based on geometry index. This code works for
+  // any combination mwms with different cross mwm sections.
+  // It's possible to implement a faster version for two special cases:
+  // * all neighboring mwms have cross_mwm section
+  // * all neighboring mwms have osrm cross mwm sections
+  TransitionPoints const points = GetTransitionPoints(s, isOutgoing);
+  for (m2::PointD const & p : points)
+  {
+    // Node. The map below is necessary because twin segments could belong to several mwm.
+    // It happens when a segment crosses more than one feature.
+    map<NumMwmId, ClosestSegment> minDistSegs;
+    auto const findBestTwins = [&](FeatureType & ft) {
```

Comment:
Yes. Thanx. Done. And we agreed to use almost equal condition instead of checking that distance is zero. It's done as well.